File: ios/sdk/swift/internal/ContactDataListener.swift
import CrossPL

/* @CrossClass */
@objc open class ContactDataListener: CrossBase {
  public enum Status: Int {
    case Unknown             = 0x0100
    case Initialized         = 0x0101
    case Destroyed           = 0x0102
    case Connecting          = 0x0104
    case Connected           = 0x0108
    case Transmitting        = 0x0110
    case Closed              = 0x0120
    case Failed              = 0x0140
  
    case PeerInitialized     = 0x1001
    case PeerFailed          = 0x1040
  }

  open func onNotify(humanCode: String, channelType: Contact.Channel, dataId: String,
                     status: Status) {
    fatalError("\(#function) not implementation.")
  }
  open func onReadData(humanCode: String, channelType: Contact.Channel, dataId: String,
                       offset: Int64, data: inout Data?) -> Int {
    fatalError("\(#function) not implementation.")
  }
  open func onWriteData(humanCode: String, channelType: Contact.Channel, dataId: String,
                        offset: Int64, data: Data?) -> Int {
    fatalError("\(#function) not implementation.")
  }

  public init() {
    super.init(className: String(describing: ContactDataListener.self))
  }
  
  /* @CrossPlatformInterface */
  @objc internal func onNotify(_ humanCode: String, _ channelType: Int, _ dataId: String,
                               _ status: Int) {
    var id = Contact.Message.FileData.ConvertId(id: dataId)
    if id == nil {
      id = dataId
    }
    
    onNotify(humanCode: humanCode, channelType: Contact.Channel(rawValue: channelType)!,
             dataId: id!, status: Status(rawValue: status)!)
  }
  
  /* @CrossPlatformInterface */
  @objc internal func onReadData(_ humanCode: String, _ channelType: Int, _ dataId: String,
                                 _ offset: Int64) -> Data? {
    var id = Contact.Message.FileData.ConvertId(id: dataId)
    if id == nil {
      id = dataId
    }
    
    var data: Data? = Data(count: 1024)
    let ret = onReadData(humanCode: humanCode, channelType: Contact.Channel(rawValue: channelType)!,
                         dataId: id!, offset: offset, data: &data)
    if(ret < 0) {
        return nil;
    }
    
    let retData = data?.subdata(in: 0..<ret)
    return retData
  }

  /* @CrossPlatformInterface */
  @objc internal func onWriteData(_ humanCode: String, _ channelType: Int, _ dataId: String,
                                  _ offset: Int64, _ data: Data?) -> Int {
    var id = Contact.Message.FileData.ConvertId(id: dataId)
    if id == nil {
      id = dataId
    }
    
    let ret = onWriteData(humanCode: humanCode, channelType: Contact.Channel(rawValue: channelType)!,
                          dataId: id!, offset: offset, data: data)
    return ret
  }
}

File: android/sdk/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactDataListener.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.nio.ByteBuffer;

@CrossClass
abstract class ContactDataListener extends CrossBase {
    public enum Status {
        Unknown             (0x0100),
        Initialized         (0x0101),
        Destroyed           (0x0102),
        Connecting          (0x0104),
        Connected           (0x0108),
        Transmitting        (0x0110),
        Closed              (0x0120),
        Failed              (0x0140),

        PeerInitialized     (0x1001),
        PeerFailed          (0x1040);

        public static Status valueOf(int id) {
            Status[] values = Status.values();
            for(int idx = 0; idx < values.length; idx++) {
                if(values[idx].id == id) {
                    return values[idx];
                }
            }
            return null;
        }

        private Status(int id){
            this.id = id;
        }
        private int id;
    }

    public abstract void onNotify(String humanCode, Contact.Channel channelType,
                                   String dataId, Status status);
    public abstract int onReadData(String humanCode, Contact.Channel channelType,
                                   String dataId, long offset, ByteBuffer data);
    public abstract int onWriteData(String humanCode, Contact.Channel channelType,
                                    String dataId, long offset, byte[] data);

    protected ContactDataListener() {
        super(ContactDataListener.class.getName(), 0);
    }

    @CrossInterface
    private void onNotify(String humanCode, int channelType,
                              String dataId, int status) {
        String id = Contact.Message.FileData.ConvertId(dataId);
        if(id == null) {
            id = dataId;
        }

        onNotify(humanCode, ContactChannel.valueOf(channelType),
                id, Status.valueOf(status));
    }

    @CrossInterface
    private byte[] onReadData(String humanCode, int channelType,
                              String dataId, long offset) {
//        Contact.Debug.DumpLocalRefTables();
        String id = Contact.Message.FileData.ConvertId(dataId);
        if(id == null) {
            id = dataId;
        }

        ByteBuffer dataBuf = ByteBuffer.allocate(1024);
        int ret = onReadData(humanCode, ContactChannel.valueOf(channelType),
                             id, offset,
                             dataBuf);
        if(ret < 0) {
            return null;
        }

        int size = ret;
        byte[] data = new byte[size];
        dataBuf.get(data, 0, size);

        return data;
    }

    @CrossInterface
    private int onWriteData(String humanCode, int channelType,
                            String dataId, long offset, byte[] data) {
//        Contact.Debug.DumpLocalRefTables();
        String id = Contact.Message.FileData.ConvertId(dataId);
        if(id == null) {
            id = dataId;
        }

        int ret = onWriteData(humanCode, ContactChannel.valueOf(channelType),
                              id, offset,
                              data);

        return ret;
    }
}
